### PR TITLE
Pruning refactoring to work with FLOPs target too

### PR DESCRIPTION
**Pull request by mkaglins**

### CODE REVIEW

File: nncf/utils.py [hunk 1]
```diff
@@ -358,3 +359,23 @@ def training_mode_switcher(model: torch.nn.Module, is_training: bool = True):
...
+    if isinstance(module, (nn.Conv1d, nn.ConvTranspose1d)):
+        ks = module.weight.data.shape
+        mac_count = ks[0] * ks[1] * ks[2] * output.shape[2]
+    elif isinstance(module, (nn.Conv2d, nn.ConvTranspose2d)):
+        ks = module.weight.data.shape
+        mac_count = ks[0] * ks[1] * ks[2] * ks[3] * output.shape[2] * output.shape[3]
+    elif isinstance(module, (nn.Conv3d, nn.ConvTranspose3d)):
+        ks = module.weight.data.shape
+        mac_count = ks[0] * ks[1] * ks[2] * ks[3] * ks[4] * output.shape[2] * \
+                    output.shape[3] * output.shape[4]
```
Comment (vanyalzr) [inline nncf/utils.py:374]: seems like you can shorten it to `np.prod(ks) * np.prod(output.shape[2:])` for convolutions of any dimension

Comment (vanyalzr) [reply nncf/utils.py:374]: also I guess you should include bias addition (in case it's present) to the FLOPs count to be exact (same for modules of `Linear` type).

**Review by vanyalzr — CHANGES_REQUESTED** [review]

File: nncf/pruning/base_algo.py [hunk 2]
```diff
@@ -331,6 +339,17 @@ def hook(grad, mask, dim=0):
                 partial_hook = update_wrapper(partial(hook, mask=mask), hook)
                 self._hooks.append(bias.register_hook(partial_hook))
 
+    def check_pruning_rate(self, params):
+        """
+        Check that set only one of pruning target params
+        """
+        pruning_target = params.get('pruning_target', None)
+        pruning_flops_target = params.get('pruning_flops_target', None)
+        if pruning_target and pruning_flops_target:
```
Comment (AlexKoff88) [inline nncf/pruning/base_algo.py:348]: What if none of the parameters is specified? We should raise an exception as well.

File: nncf/utils.py [hunk 3]
```diff
@@ -358,3 +359,23 @@ def training_mode_switcher(model: torch.nn.Module, is_training: bool = True):
...
+    elif isinstance(module, nn.Linear):
+        mac_count = input_[0].shape[1] * output.shape[-1]
+    elif isinstance(module, (nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d)):
```
Comment (AlexKoff88) [inline nncf/utils.py:377]: We should not count BatchNorms because they are fused at inference time.

File: nncf/pruning/filter_pruning/algo.py [hunk 4]
```diff
@@ -83,19 +98,178 @@ def statistics(self, quickly_collected_only=False):
     def freeze(self):
         self.frozen = True
 
+    def _init_pruned_modules_params(self):
+        def get_in_out_channels(module):
+            in_channels, out_channels = None, None
+            if isinstance(module, (nn.Conv1d, nn.Conv2d, nn.Conv3d, nn.ConvTranspose1d, nn.ConvTranspose2d,
```
Comment (AlexKoff88) [inline nncf/pruning/filter_pruning/algo.py:104]: It seems you already have this code in utils.py. I propose to make it more granular and reuse here.

File: nncf/pruning/filter_pruning/algo.py [hunk 5]
```diff
@@ -83,19 +98,178 @@ def statistics(self, quickly_collected_only=False):
...
+    def flops_count_init(self):
+        def get_node_flops_hook(name, dict_to_save):
+            return partial(compute_FLOPs_hook, dict_to_save=dict_to_save, name=name)
+
+        def get_node_cost_hook(name):
+            """
+            Cost of node is num of flops for this node divided by numbers of input and output channels for this node.
+            """
+
+            def compute_cost_hook(module, input_, output):
```
Comment (AlexKoff88) [inline nncf/pruning/filter_pruning/algo.py:147]: The same

File: nncf/pruning/base_algo.py (hunk 2 above)
Comment (mkaglins) [reply nncf/pruning/base_algo.py:348]: @AlexKoff88 Why? We had default `pruning_target` = 0.5.

Comment (AlexKoff88) [reply nncf/pruning/base_algo.py:348]: Ok. I found that in the code below.

File: nncf/utils.py (hunk 1 above)
Comment (mkaglins) [reply nncf/utils.py:374]: The function was shortened and bias addition was added.

File: nncf/pruning/filter_pruning/algo.py (hunk 4 above)
Comment (mkaglins) [reply nncf/pruning/filter_pruning/algo.py:104]: These functions have clear objectives and aren't similar enough to be united. I propose to leave them separately to keep them simple.

File: nncf/pruning/filter_pruning/algo.py (hunk 5 above)
Comment (mkaglins) [reply nncf/pruning/filter_pruning/algo.py:144]: Same comment as above

File: nncf/utils.py (hunk 3 above)
Comment (mkaglins) [reply nncf/utils.py]: BatchNom is not taking into account in FLOPs calculation anymore.

File: nncf/pruning/filter_pruning/algo.py [hunk 6]
```diff
@@ -83,19 +98,170 @@ def statistics(self, quickly_collected_only=False):
...
+                tmp_out_channels[node.nncf_node_id] = new_out_channels_num
+
+            # Prune in_channels in all next nodes of cluster
+            next_nodes = self.next_nodes[group.id]
+            for node_id in next_nodes:
+                tmp_in_channels[node_id] = new_out_channels_num
+        flops = self._calculate_flops_in_pruned_model(tmp_in_channels, tmp_out_channels)
+        return flops
+
+    def _find_layerwise_pruning_rate(self, target_flops_pruning_rate):
```
Comment (mkaglins) [inline nncf/pruning/filter_pruning/algo.py:221]: @AlexKoff88 Please, take a closer look at this algo of layer-wise pruning rate search for FLOPs pruning target (quickly discussed earlier).

Comment (AlexKoff88) [reply nncf/pruning/filter_pruning/algo.py:221]: LGTM

**Review by AlexKoff88 — APPROVED** [review]

File: tests/quantization/test_precision_init.py [hunk 7]
```diff
@@ -711,7 +711,8 @@ def get_quantization_config_with_ignored_scope():
 @pytest.mark.parametrize(('config_creator', 'ref_values'), (
     [
         get_quantization_config_without_range_init,
-        (1.75, pytest.approx(1.07, abs=1e-2), (1, 2), (1, 4), (1, 1.12))
+        (pytest.approx(1.57, abs=1e-2), pytest.approx(1.15, abs=1e-2), (1, 2), (1, 4),
```
Comment (mkaglins) [inline tests/quantization/test_precision_init.py:714]: @ljaljushkin I deleted BatchNorms from FLOPs calculating and fix test in quantization used FLOPs count. Please, check that it is correct now.

**Comment by mkaglins** [comment]
@vanyalzr Please, verify that all requested changes are done.

File: tests/quantization/test_precision_init.py (hunk 7 above)
Comment (ljaljushkin) [reply tests/quantization/test_precision_init.py:714]: It's strange that FLOPS were changed after deleting BN because the test model doesn't have it.
Looks like there's a change in the formula for Conv and Linear

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/quantization/test_precision_init.py (hunk 7 above)
Comment (ljaljushkin) [reply tests/quantization/test_precision_init.py:714]: Could you verify that's discrepancy happens because of considering `bias`?
```
if module.bias is not None:
            flops_count += np.prod(output.shape[1:])
```

Comment (mkaglins) [reply tests/quantization/test_precision_init.py:714]: @ljaljushkin For Conv and Linear bias addition was added to FLOPs calculation. Check it in `nncf/utils.py`

Comment (mkaglins) [reply tests/quantization/test_precision_init.py:714]: Ok, I will re-check it.

Comment (AlexKoff88) [reply tests/quantization/test_precision_init.py:714]: Biased of Conv and FC should be also excluded from FLOPS

Comment (mkaglins) [reply tests/quantization/test_precision_init.py]: OK, biases are deleted from FLOPs calculation and tests for quantization now work without fixes.

**Review by ljaljushkin — APPROVED** [review]

**Comment by AlexKoff88** [comment]
Are there any leftovers that really block us from merging this?

**Comment by mkaglins** [comment]
No, just formal approval from @vanyalzr

**Review by vanyalzr — APPROVED** [review]

**Comment by mkaglins** [comment]
@vshampor Please, merge this PR. All reviews are done.
